src/apps/lwaftr/ipv6_apps.lua

```diff
@@ -71,15 +75,37 @@ end
 
 function Reassembler:cache_fragment(frag)
    local cache = self.fragment_cache
+
+   if self.fragment_count >= self.fragment_cache_max then
+      local n = math.random(self.fragment_key_count)
+      for _, frags in pairs(cache) do
```

Comment:
I have been thinking about this, and I believe to have found the best balance between simplicity and keeping all the operations O(1). Each fragment list stored in the cache would get the following additional fields:

- A `.time` field, with the time at which a packet was last added to it. It gets updated every time a fragment is added to that particular fragment list.
- `.next` and `.prev` fields, used to chain all fragment lists in a doubly-linked list. The list is kept sorted by `.time` at all times.

On top of the existing `cache` table which maps keys to fragment lists, the doubly linked list is updated whenever a new fragment arrives:

1. If the fragment list to which it belongs already exists, it is removed from the linked list (this only needs assigning `.next` and `.prev` “pointers”, it's a constant-time operation). Otherwise a new fragment list is created and addded
1. The current time is set as the `.time` of the fragment list.
1. The fragment is appended to the fragment list (also, “pointer” fiddling: constant time).
1. The fragment list is _appended_ to the fragment list.

As for trimming fragments, it can be done by traversing the linked list starting from the end. Periodic trimming of old fragments can be done in `Reassembler:push()`, as long as there is a limit in the maximum amount of elements of the linked list checked each time—to make sure that trimming does not consume more than a maximum fixed time of CPU.